File: src/main/java/org/antlr/intellij/plugin/listeners/PreViewListener.java
package org.antlr.intellij.plugin.listeners;

import com.intellij.openapi.editor.event.EditorMouseEvent;
import com.intellij.openapi.vfs.VirtualFile;
import org.antlr.intellij.plugin.preview.PreviewState;
import org.jetbrains.annotations.Nullable;

import java.util.EventListener;

public interface PreViewListener extends EventListener {
    void releaseEditor(PreviewState previewState);

    void setStartRuleName(VirtualFile grammarFile, String startRuleName);

    void updateParseTreeFromDoc(VirtualFile grammarFile);

    void grammarFileSaved(VirtualFile grammarFile);

    void grammarFileChanged(VirtualFile grammarFile);

    void mouseEnteredGrammarEditorEvent(VirtualFile file, EditorMouseEvent event);

    void closeGrammar(VirtualFile file);

    void setEnabled(boolean enabled);

    void toolWindowHide(@Nullable Runnable runnable);

    void onParsingCompleted(PreviewState previewState, long duration);

    void notifySlowParsing();

    void onParsingCancelled();

    void clearParseErrors();

    void startParsing();

    void autoRefreshPreview(VirtualFile virtualFile);

}

File: src/main/java/com/antlr/listener/PreViewListener.kt
package com.antlr.listener

import com.antlr.preview.PreviewState
import com.antlr.ui.view.PreviewPanel
import com.intellij.openapi.vfs.VirtualFile

class PreViewListener(val previewPanel: PreviewPanel) : AbstractAntlrListener() {
    override fun releaseEditor(previewState: PreviewState) {
        previewPanel.getInputPanel().releaseEditor(previewState)
    }

    override fun startRuleName(virtualFile: VirtualFile, ruleName: String) {
        previewPanel.getInputPanel().setStartRuleName(virtualFile, ruleName)
    }

    override fun updateParseTreeState(virtualFile: VirtualFile) {
        previewPanel.updateParseTreeFromDoc(virtualFile)
    }

    override fun grammarChange(newFile: VirtualFile) {
        previewPanel.grammarFileChanged(newFile)
    }

    override fun grammarSaved(virtualFile: VirtualFile) {
        previewPanel.grammarFileSaved(virtualFile)
    }

    override fun parsingCompleted(previewState: PreviewState, ms: Long) {
        previewPanel.onParsingCompleted(previewState, ms)
    }

    override fun parsingSlow() {
        previewPanel.notifySlowParsing()
    }

    override fun cancelParsing() {
        previewPanel.onParsingCancelled()
    }

    override fun clearParseErrors() {
        previewPanel.inputPanel.clearParseErrors()
    }

    override fun startParsing() {
        previewPanel.startParsing()
    }

    override fun autoRefreshPreview(virtualFile: VirtualFile) {
        previewPanel.autoRefreshPreview(virtualFile)
    }
}
